packages/core/admin/server/services/api-token.js

```diff
@@ -1,30 +1,106 @@
 'use strict';
 
 const crypto = require('crypto');
+const { omit, difference, isEmpty, map, isArray } = require('lodash/fp');
+const { ValidationError, NotFoundError } = require('@strapi/utils').errors;
+const constants = require('./constants');
 
 /**
- * @typedef {'read-only'|'full-access'} TokenType
+ * @typedef {'read-only'|'full-access'|'custom'} TokenType
  */
 
 /**
  * @typedef ApiToken
  *
  * @property {number|string} id
  * @property {string} name
- * @property {string} [description]
+ * @property {string} description
  * @property {string} accessKey
+ * @property {number} lastUsedAt
  * @property {TokenType} type
+ * @property {(number|ApiTokenPermission)[]} permissions
  */
 
+/**
+ * @typedef ApiTokenPermission
+ *
+ * @property {number|string} id
+ * @property {string} action
+ * @property {ApiToken|number} token
+ */
+
+/** @constant {Array<string>} */
+const SELECT_FIELDS = ['id', 'name', 'description', 'lastUsedAt', 'type', 'createdAt', 'updatedAt'];
+
 /** @constant {Array<string>} */
-const SELECT_FIELDS = ['id', 'name', 'description', 'type', 'createdAt'];
+const POPULATE_FIELDS = ['permissions'];
+
+/**
+ * Assert that a token's permissions attribute is valid for its type
+ *
+ * @param {ApiToken} token
+ */
+const assertCustomTokenPermissionsValidity = (attributes) => {
+  // Ensure non-custom tokens doesn't have permissions
+  if (attributes.type !== constants.API_TOKEN_TYPE.CUSTOM && !isEmpty(attributes.permissions)) {
+    throw new ValidationError('Non-custom tokens should not reference permissions');
+  }
+
+  // Custom type tokens should always have permissions attached to them
+  if (attributes.type === constants.API_TOKEN_TYPE.CUSTOM && isEmpty(attributes.permissions)) {
+    throw new ValidationError('Missing permissions attribute for custom token');
+  }
+};
+
+/**
+ * Flatten a token's database permissions objects to an array of strings
+ *
+ * @param {ApiToken} token
+ *
+ * @returns {ApiToken}
+ */
+const flattenTokenPermissions = (token) => {
+  if (!token) return token;
+  return {
+    ...token,
+    permissions: isArray(token.permissions) ? map('action', token.permissions) : token.permissions,
+  };
+};
 
 /**
+ *  Get a token
+ *
+ * @param {Object} whereParams
+ * @param {string|number} whereParams.id
+ * @param {string} whereParams.name
+ * @param {number} whereParams.lastUsedAt
+ * @param {string} whereParams.description
+ * @param {string} whereParams.accessKey
+ *
+ * @returns {Promise<Omit<ApiToken, 'accessKey'> | null>}
+ */
+const getBy = async (whereParams = {}) => {
+  if (Object.keys(whereParams).length === 0) {
+    return null;
+  }
+
+  const token = await strapi
+    .query('admin::api-token')
+    .findOne({ select: SELECT_FIELDS, populate: POPULATE_FIELDS, where: whereParams });
+
+  if (!token) return token;
+  return flattenTokenPermissions(token);
```

Comment:
Does it make sense to flatten everytime?
If a controller uses this call to get back a token + populate its permissions, then try to return this data as a basic content-api response, it'll fail to map the token-permission's attributes to the actual data.

Maybe we could decouple the flattening of the result from this method. What do you think?